cmd/ctlog/verifyfulcio/main.go

```diff
@@ -0,0 +1,136 @@
+	"github.com/sigstore/sigstore/pkg/cryptoutils"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/signals"
+)
+
+type fulcios []string
+
+func (f *fulcios) String() string {
+	return fmt.Sprint(*f)
+}
+
+func (f *fulcios) Set(value string) error {
+	*f = append(*f, value)
+	return nil
+}
+
+var fulcioList fulcios
+
+type CertResponse struct {
+	Certificates []string `json:"certificates"`
+}
+
+func main() {
+
+	flag.Var(&fulcioList, "fulcio", "List of fulcios which must be in the list")
+	var ctlogURL = flag.String("ctlog-url", "ctlog.ctlog-system.svc", "CTLog to check Fulcios against.")
+	var ctlogPrefix = flag.String("log-prefix", "sigstorescaffolding", "Prefix to append to the gtlogURL url. This is basically the name of the log.")
+	flag.Parse()
+	var strictMatch = flag.Bool("strict", true, "If set to true ctlog must only contain the Fulcios in the list, no more, no less")
+	ctx := signals.NewContext()
+	fulcioURLs := make([]*url.URL, 0, len(fulcioList))
+	for _, f := range fulcioList {
+		u, err := url.Parse(f)
+		if err != nil {
+			logging.FromContext(ctx).Fatalf("Invalid fulcioURL %s : %v", f, err)
+		}
+		fulcioURLs = append(fulcioURLs, u)
+	}
+	fmt.Printf("GOT: %s\n", fulcioList.String())
+
+	// First grab the certs that CTLog has.
+	ctlog := fmt.Sprintf("%s/%s/ct/v1/get-roots", *ctlogURL, *ctlogPrefix)
+	ctlogResponse, err := http.Get(ctlog)
+	if err != nil {
+		logging.FromContext(ctx).Fatalf("Failed to get trusted certs from ctlog: %v", err)
+	}
+	defer ctlogResponse.Body.Close()
+	body, err := io.ReadAll(ctlogResponse.Body)
+	if err != nil {
+		logging.FromContext(ctx).Fatalf("Failed to read body from ctlog: %v", err)
+	}
+	certs := CertResponse{}
+	if err := json.Unmarshal(body, &certs); err != nil {
+		logging.FromContext(ctx).Fatalf("Failed to unmarshal body from ctlog: %v", err)
+	}
+	for i, cert := range certs.Certificates {
+		logging.FromContext(ctx).Infof("Got back cert %d: %s", i, cert)
+	}
+
+	// Keep track of certs found. Same index as fulcioURLs
+	certsFound := make([]bool, len(fulcioURLs))
+	for i, fulcio := range fulcioURLs {
+		logging.FromContext(ctx).Infof("Fetching fulcio cert %s", fulcio)
+		client := fulcioclient.NewClient(fulcio)
+		root, err := client.RootCert()
+		if err != nil {
+			logging.FromContext(ctx).Fatalf("Failed to fetch fulcio Root %s cert: %v", fulcio.String(), err)
+		}
+		fulcioCerts, err := cryptoutils.UnmarshalCertificatesFromPEM(root.ChainPEM)
+		if err != nil {
+			logging.FromContext(ctx).Fatalf("Failed to unmarshal fulcio Root cert %s cert: %v", fulcio.String(), err)
+		}
+		fulcioRoot, err := cryptoutils.MarshalCertificateToPEM(fulcioCerts[len(fulcioCerts)-1])
+		if err != nil {
+			logging.FromContext(ctx).Fatalf("Failed to marshal fulcio Root cert %s cert: %w", fulcio.String(), err)
+		}
+		logging.FromContext(ctx).Infof("Got a root cert for fulcio Root cert %s", fulcio)
+		// Strip the certificate begin/end marker strings since CTLog doesn't
```

Comment:
Is this accurate? https://github.com/sigstore/public-good-instance/blob/main/config/ctfe/root.pem shows that we did specify PEM headers. Or is this just how we store them in the ctlog config, and then add headers back later on?

Reply:
This is how they are being served from the logs:
```
vaikas@villes-mbp scaffolding % curl https://ctfe.sigstore.dev/test/ct/v1/get-roots | jq .
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1381  100  1381    0     0   4538      0 --:--:-- --:--:-- --:--:--  4681
{
  "certificates": [
    "MIIB+DCCAX6gAwIBAgITNVkDZoCiofPDsy7dfm6geLbuhzAKBggqhkjOPQQDAzAqMRUwEwYDVQQKEwxzaWdzdG9yZS5kZXYxETAPBgNVBAMTCHNpZ3N0b3JlMB4XDTIxMDMwNzAzMjAyOVoXDTMxMDIyMzAzMjAyOVowKjEVMBMGA1UEChMMc2lnc3RvcmUuZGV2MREwDwYDVQQDEwhzaWdzdG9yZTB2MBAGByqGSM49AgEGBSuBBAAiA2IABLSyA7Ii5k+pNO8ZEWY0ylemWDowOkNa3kL+GZE5Z5GWehL9/A9bRNA3RbrsZ5i0JcastaRL7Sp5fp/jD5dxqc/UdTVnlvS16an+2Yfswe/QuLolRUCrcOE2+2iA5+tzd6NmMGQwDgYDVR0PAQH/BAQDAgEGMBIGA1UdEwEB/wQIMAYBAf8CAQEwHQYDVR0OBBYEFMjFHQBBmiQpMlEk6w2uSu1KBtPsMB8GA1UdIwQYMBaAFMjFHQBBmiQpMlEk6w2uSu1KBtPsMAoGCCqGSM49BAMDA2gAMGUCMH8liWJfMui6vXXBhjDgY4MwslmN/TJxVe/83WrFomwmNf056y1X48F9c4m3a3ozXAIxAKjRay5/aj/jsKKGIkmQatjI8uupHr/+CxFvaJWmpYqNkLDGRU+9orzh5hI2RrcuaQ==",
    "MIIB9zCCAXygAwIBAgIUALZNAPFdxHPwjeDloDwyYChAO/4wCgYIKoZIzj0EAwMwKjEVMBMGA1UEChMMc2lnc3RvcmUuZGV2MREwDwYDVQQDEwhzaWdzdG9yZTAeFw0yMTEwMDcxMzU2NTlaFw0zMTEwMDUxMzU2NThaMCoxFTATBgNVBAoTDHNpZ3N0b3JlLmRldjERMA8GA1UEAxMIc2lnc3RvcmUwdjAQBgcqhkjOPQIBBgUrgQQAIgNiAAT7XeFT4rb3PQGwS4IajtLk3/OlnpgangaBclYpsYBr5i+4ynB07ceb3LP0OIOZdxexX69c5iVuyJRQ+Hz05yi+UF3uBWAlHpiS5sh0+H2GHE7SXrk1EC5m1Tr19L9gg92jYzBhMA4GA1UdDwEB/wQEAwIBBjAPBgNVHRMBAf8EBTADAQH/MB0GA1UdDgQWBBRYwB5fkUWlZql6zJChkyLQKsXF+jAfBgNVHSMEGDAWgBRYwB5fkUWlZql6zJChkyLQKsXF+jAKBggqhkjOPQQDAwNpADBmAjEAj1nHeXZp+13NWBNa+EDsDP8G1WWg1tCMWP/WHPqpaVo0jhsweNFZgSs0eE7wYI4qAjEA2WB9ot98sIkoF3vZYdd3/VtWB5b9TNMea7Ix/stJ5TfcLLeABLE4BNJOsQ4vnBHJ"
  ]
}
```

Reply:
Ah interesting, this is base64 encoded DER. I guess `pem.Decode` handles headerless PEM?